### PR TITLE
feat: Modern account switcher

### PR DESCRIPTION
**Pull request by adrien-coye**
- Double tap on the avatar should bring up the next account
- Long touch should display the account list

Double tap looks like a hack because it is one, as tap events over a tabBar are not something iOS allows you to do easily.
Yet this implementation is relatively safe.

~~Blocked by for `AppNavigable` (router class) https://github.com/Infomaniak/ios-kDrive/pull/1187~~

https://github.com/Infomaniak/ios-kDrive/assets/121806582/0da271d8-fffb-46fe-bc19-469779b0051d

### CODE REVIEW

File: kDrive/UI/Controller/MainTabViewController.swift [hunk 1]
```diff
@@ -251,13 +256,31 @@ extension MainTabViewController: MainTabBarDelegate {
...
+        let topViewController = navigationController.topViewController
+        if let homeViewController = topViewController as? HomeViewController {
             homeViewController.presentedFromTabBar()
         }
 
-        if tabBarController.selectedViewController == viewController,
-           let viewController = (viewController as? UINavigationController)?.topViewController as? TopScrollable {
-            viewController.scrollToTop()
+        if tabBarController.selectedViewController == viewController {
+            // Detect double tap on menu
```
Comment (adrien-coye) [inline kDrive/UI/Controller/MainTabViewController.swift:273]: Added a comment cuz it's a hack

File: kDrive/UI/View/Menu/MainTabBar.swift [hunk 2]
```diff
@@ -163,6 +165,30 @@ final class MainTabBar: UITabBar {
...
+        }
+
+        // Touch is over the 5th button's x position
+        let touchPoint = recognizer.location(in: self)
+        guard touchPoint.x > bounds.width / 5 * 4 else {
+            return
+        }
+
+        let generator = UIImpactFeedbackGenerator(style: .light)
+        generator.impactOccurred()
```
Comment (adrien-coye) [inline kDrive/UI/View/Menu/MainTabBar.swift:186]: A `light` impact felt more refined than a medium or a heavy on iP8.

I also think it adds to the experience on the long touch only

File: kDrive/UI/View/Menu/MainTabBar.swift [hunk 3]
```diff
@@ -163,6 +165,30 @@ final class MainTabBar: UITabBar {
...
+
+        print("Long touch")
+        // TODO: Nav to detail View, wait for AppNavigable to be available
```
Comment (adrien-coye) [inline kDrive/UI/View/Menu/MainTabBar.swift:189]: I'll finish this once https://github.com/Infomaniak/ios-kDrive/pull/1187 is merged

File: kDrive/UI/View/Menu/MainTabBar.swift [hunk 4]
```diff
@@ -163,6 +165,30 @@ final class MainTabBar: UITabBar {
...
+        addGestureRecognizer(longTouch)
+    }
+
+    @objc func handleLongTouch(recognizer: UITapGestureRecognizer) {
+        guard recognizer.state == .began else {
+            return
+        }
+
+        // Touch is over the 5th button's x position
+        let touchPoint = recognizer.location(in: self)
```
Comment (adrien-coye) [inline kDrive/UI/View/Menu/MainTabBar.swift:180]: Added a comment cuz it's a hack also

File: kDrive/UI/View/Menu/MainTabBar.swift [hunk 5]
```diff
@@ -16,6 +16,7 @@
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+import InfomaniakDI
```
Comment (adrien-coye) [inline kDrive/UI/View/Menu/MainTabBar.swift:19]: Will be used for the navigation code as soon as available

File: kDrive/UI/Controller/MainTabViewController.swift [hunk 6]
```diff
@@ -251,13 +256,31 @@ extension MainTabViewController: MainTabBarDelegate {
...
+            if let viewController = topViewController as? MenuViewController,
+               let lastDate = lastInteraction,
+               Date().timeIntervalSince(lastDate) <= Self.doubleTapInterval {
+                print("double tap detected")
```
Comment (adrien-coye) [inline kDrive/UI/Controller/MainTabViewController.swift:277]: I'll finish this once https://github.com/Infomaniak/ios-kDrive/pull/1187 is merged

**Review by adrien-coye — COMMENTED** [review]
Left some notes for furure reviewers

File: kDrive/UI/Controller/MainTabViewController.swift [hunk 7]
```diff
@@ -243,19 +250,66 @@ extension MainTabViewController: MainTabBarDelegate {
...
+    func avatarDoubleTap() {
+        accountManager.switchToNextAvailableAccount()
+        guard let accountManager = accountManager.currentDriveFileManager else {
+            return
+        }
+
+        let generator = UIImpactFeedbackGenerator(style: .light)
+        generator.impactOccurred()
+
+        @InjectService var router: AppNavigable
```
Comment (valentinperignon) [inline kDrive/UI/Controller/MainTabViewController.swift:278]: Do we need to import the router here as it's already a property of MainTabViewController?

File: kDrive/UI/Controller/MainTabViewController.swift [hunk 8]
```diff
@@ -243,19 +250,66 @@ extension MainTabViewController: MainTabBarDelegate {
         floatingPanelViewController.trackAndObserve(scrollView: plusButtonFloatingPanel.tableView)
         present(floatingPanelViewController, animated: true)
     }
+
+    func avatarLongTouch() {
+        let viewControllers = viewControllers
```
Comment (valentinperignon) [inline kDrive/UI/Controller/MainTabViewController.swift:255]: Why do we need to create a second variable `viewControllers`?

File: kDrive/UI/View/Menu/MainTabBar.swift (hunk 5 above)
Comment (valentinperignon) [reply kDrive/UI/View/Menu/MainTabBar.swift:19]: Why do wee need to add an import that is useless for the moment?

File: kDrive/UI/Controller/MainTabViewController.swift (hunk 7 above)
Comment (adrien-coye) [reply kDrive/UI/Controller/MainTabViewController.swift:278]: Good catch, that's a mistake.

File: kDrive/UI/Controller/MainTabViewController.swift (hunk 8 above)
Comment (adrien-coye) [reply kDrive/UI/Controller/MainTabViewController.swift:255]: I do not know, there is no good reason for it. It is probably something that I did not clean up after the PoC stage.

**Review by valentinperignon — APPROVED** [review]

**Pull request merged**